---
title: "NGS Analysis"
author: "Andrea Ouyang"
date: "10/28/2020"
output: html_document
---

```{r}
setwd("~/Desktop/bioinfo/assn1")
library("DESeq2")
library("pheatmap")
library("RColorBrewer")
#library("vsn")
library("AnnotationDbi")
library("org.Hs.eg.db")
library("genefilter")
library("biomaRt")
#library("IHW")
library("ggplot2")
library("gplots")
library(Rsubread)
```
Did the featureCounts in cluster, moved readcounts into own space
```{r}
readcounts = featureCounts(dir(pattern = "sorted.bam"),                         annot.ext="Homo_sapiens.GRCh38.101.gtf", isGTFAnnotationFile=T, countMultiMappingReads=F)
save.image(file="1.RData")
#moved .RData to local environment
```

Preliminary analysis using DESeq2: preprocessing
```{r}
load("/Users/andreaouy/Desktop/bioinfo/assn1/1.RData")
head(readcounts)
expdesign = factor(c("exp", "cont", "cont", "cont", "exp", "exp"))
counts=readcounts$counts
```

Dispersion estimates
```{r}
cds <- DESeqDataSetFromMatrix(countData=counts,
                             colData=data.frame(condition=expdesign),
                             design= ~ condition)
cds <- estimateSizeFactors(cds) #consider the different reads in total reads from different samples
cds <- estimateDispersions(cds)#estimate the dispersion/variation between replicates
plotDispEsts(cds) #as the read count increases, dispersion decreases
```

```{r}
# First we transform our raw count data using a variance stabilizing transformation (VST) that roughly mirrors how DeSeq2 models the data.
vsd1 <- varianceStabilizingTransformation(cds, blind=FALSE)
# Then we plot a PCA, grouping and coloring our datasets according to batch
plotPCA(vsd1, "condition") # looks separated, biological replicates grouped together
vsd2 <- varianceStabilizingTransformation(cds, blind=FALSE)
assay(vsd2) <- limma::removeBatchEffect(assay(vsd2), vsd2$batch)
plotPCA(vsd2, "condition") #looks the same as the first - batch effect correction is fine

sampleDistsCorr <- dist( t( assay(vsd2) ) )
sampleDistsCorr
sampleDistCorrMatrix <- as.matrix( sampleDistsCorr )
colnames(sampleDistCorrMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistCorrMatrix, clustering_distance_rows=sampleDistsCorr, clustering_distance_cols=sampleDistsCorr, col=colors)
#fantastic! grouped in the way we wanted it to
```

```{r}
#DESeq2 negative binomial normalization
cds <- DESeq(cds) #will spit out what we want to compare, then use adjusted p-value as cutoff
res <- results(cds)
head(res)
class(res$padj) #class is numeric
#we select the p-adjusted values from res that are less than 0.001, the sum function will 
#add up the number of "T" in the vetor. Remove the NA values
sum(res$padj < 0.05, na.rm=T) #10408 elements; padj is arbitrary
plotMA(res)
```

```{r}
baseMeanCtrl = rowMeans(counts(cds,normalized=TRUE)[,cds$condition == "cont"])
baseMeanExp = rowMeans(counts(cds,normalized=TRUE)[,cds$condition == "exp"])
res1 <-  cbind(as.data.frame(counts), baseMeanCtrl, baseMeanExp)
res1$log2FoldChange <- res$log2FoldChange
res1$pvalue <- res$pvalue
res1$padj <- res$padj

# Here we add two further columns, the gene symbol (common name) and entrez ID - both of which may be useful downstream
res1$symbol <- mapIds(org.Hs.eg.db, keys=row.names(res1), column="SYMBOL", keytype="ENSEMBL", multiVals="first") # MAPS GENE IDs
res1$entrez <- mapIds(org.Hs.eg.db, keys=row.names(res1), column="ENTREZID", keytype="ENSEMBL", multiVals="first")

# Finally we write the complete results object to an outfile
write.csv(res1, "DGEanalysis.csv", row.names=TRUE)
```

Plotting

```{r}
#library(EnhancedVolcano)
EnhancedVolcano(toptable = res1, x = "log2FoldChange", y = "pvalue", 
lab = res1$symbol, pCutoff = 0.01, FCcutoff = 1, 
xlim = c(-5, 5), ylim = c(-20, 350), 
title = "featureCounts DESeq2 results", 
subtitle = "Differential Expression Analysis", caption = NULL, 
col = c("grey30", "forestgreen", "purple", "dark orange"))
```

```{r}
resSigind <- res[ which(res$padj < 0.01 & res$log2FoldChange > 0), ] #which genes were induced
resSigrep <- res[ which(res$padj < 0.01 & res$log2FoldChange < 0), ] #which genes were repressed
resSig <- rbind(resSigind, resSigrep)
normvalues <- counts(cds, normalize = T)
sigvalues <- normvalues[resSig@rownames, ]

counts.filt.cor <- cor(t(sigvalues)) 
counts.filt.dist <- as.dist(1-counts.filt.cor)#converted to distances
counts.filt.hclust <- hclust(counts.filt.dist, method = "ave")
plot(counts.filt.hclust, hang = -1, cex = 0.1, main = "DESeq2 Cluster") #obviously 2 clusters
```

```{r}
# top 20 variable genes
logfold <- resSig[order(abs(resSig$log2FoldChange)),]
top20 <- logfold[1:20,]
rld <- rlog(cds) # apply "regularized log" transformation to DESeqDataSet
df <- as.data.frame(colData(cds)) 
mat <- assay(rld)[rownames(top20), ] # generate new matrix w/top20 variable genes
mat<-mat-rowMeans(mat) ##plot distances from the mean, makes heatmap clearer
mart <- useMart("ensembl","hsapiens_gene_ensembl") ## assuming human
gns <- getBM(c("hgnc_symbol","ensembl_gene_id"), "ensembl_gene_id", row.names(mat), mart)
row.names(mat)[match(gns[,2], row.names(mat))] <- gns[,1]
df <- as.data.frame(colData(cds))
pheatmap(mat, cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df, cex=0.9, main= "Top 20 Variable Genes")

```

```{r}
ordered <- resSig[order(resSig$padj), ] 
subset.top <- head((subset(ordered, ordered$padj < 0.01)), n=25) 
sigGenes <- rownames(subset.top)
rows <- match(sigGenes, row.names(rld))
mat <- assay(rld)[rows,] # to extract matrix 
mat <- mat-rowMeans(mat) ## plot distances from the mean, makes heatmap clearer
mart <- useMart("ensembl","hsapiens_gene_ensembl") ## assuming human
gns <- getBM(c("hgnc_symbol","ensembl_gene_id"), "ensembl_gene_id", row.names(mat), mart)
row.names(mat)[match(gns[,2], row.names(mat))] <- gns[,1]
df <- as.data.frame(colData(cds))
pheatmap(mat, cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df, main = "featureCounts - Top 25 most significant genes", cex=0.8)
```









